### PR TITLE
Changed type of the 'type' field to foreign key to ContentType

### PR DESCRIPTION
**Pull request by IO5**
@ https://github.com/craigds/django-typed-models/issues/56

> At the same time perhaps we should change it to be a content-type foreign key rather than a app.model string. So the actual db column would be dtm_type_id I guess.

I decided to have a go at it. It "works for me" i.e. in my hobby project (full disclosure: I'm not a web developer by trade), but it still does require more testing.

It should be a good starting point if nothing else.

I had to do a custom migration, which was easy enough, but I wonder if there is a way for a package to generate it automatically?
It's basically:
```python
    app_label, model = object.type.split('.')
    object.dtm_type_id = ContentType.objects.get(app_label=app_label, model=model).id
    object.save()
```

### CODE REVIEW

File: typedmodels/models.py [hunk 1]
```diff
@@ -410,21 +410,21 @@ def recast(self, typ=None):
...
+                self.dtm_type = self._typedmodels_type
 
         if typ is None:
-            typ = self.type
+            typ = self.dtm_type
         else:
             if isinstance(typ, type) and issubclass(typ, base):
                 model_name = typ._meta.model_name
-                typ = "%s.%s" % (typ._meta.app_label, model_name)
+                typ = ContentType.objects.get(app_label=typ._meta.app_label, model=model_name)
```
Comment (IO5) [inline typedmodels/models.py:420]: This line is a little iffy, it can probably be done without querying the DB

Comment (craigds) [reply typedmodels/models.py:420]: https://docs.djangoproject.com/en/3.2/ref/contrib/contenttypes/#django.contrib.contenttypes.models.ContentTypeManager.get_for_model

```suggestion
                typ = ContentType.objects.get_for_model(typ)
```

File: typedmodels/models.py [hunk 2]
```diff
@@ -160,20 +162,18 @@ def do_related_class(self, other, cls):
             opts = cls._meta
 
             model_name = opts.model_name
-            typ = "%s.%s" % (opts.app_label, model_name)
+            typ = ContentType.objects.get(app_label=opts.app_label, model=model_name)
```
Comment (craigds) [inline typedmodels/models.py:165]:
```suggestion
            typ = ContentType.objects.get_for_model(cls)
```

File: typedmodels/models.py [hunk 3]
```diff
@@ -205,15 +205,15 @@ def get_type_classes(subcls):
...
+                        cls._typedmodels_registry[k.id]
                         for k in subcls._typedmodels_subtypes
                     ]
 
             cls.get_type_classes = classmethod(get_type_classes)
 
             def get_types(subcls):
                 if subcls is cls:
-                    return list(cls._typedmodels_registry.keys())
+                    return ContentType.objects.get(id__in=list(cls._typedmodels_registry.keys()))
```
Comment (craigds) [inline typedmodels/models.py:216]: This looks incorrect - `.get()` returns a single object but we want a list.

I'd suggest something like this instead:
```suggestion
                    return [ContentType.objects.get_for_id(pk) for pk in cls._typedmodels_registry.keys()]
```

File: typedmodels/models.py [hunk 4]
```diff
@@ -321,8 +321,8 @@ def say_something(self):
 
     objects = TypedModelManager()
 
-    type = models.CharField(
-        choices=(), max_length=255, null=False, blank=False, db_index=True
+    dtm_type = models.ForeignKey(
+        ContentType, null=True, blank=False, db_index=True, on_delete=models.PROTECT
```
Comment (craigds) [inline typedmodels/models.py:325]: Why the change to `null=True`?

File: typedmodels/models.py [hunk 5]
```diff
@@ -335,14 +335,14 @@ class Meta:
     def from_db(cls, db, field_names, values):
         # Called when django instantiates a model class from a queryset.
         _typedmodels_do_recast = True
-        if "type" not in field_names:
+        if "dtm_type_id" not in field_names:
             # LIMITATION:
             # `type` was deferred in the queryset.
```
Comment (craigds) [inline typedmodels/models.py:340]:
```suggestion
            # `dtm_type` was deferred in the queryset.
```

File: typedmodels/models.py [hunk 6]
```diff
@@ -400,7 +400,7 @@ def recast(self, typ=None):
         else:
             raise ValueError("No suitable base class found to recast!")
 
-        if not self.type:
+        if not self.dtm_type:
```
Comment (craigds) [inline typedmodels/models.py:403]: Unless django is doing something magic with contenttype foreignkeys (I don't think so) this will do a db query to fetch the ContentType object. When all we really need to check is the local attribute:

```suggestion
        if self.dtm_type_id is None:
```

File: typedmodels/models.py [hunk 7]
```diff
@@ -410,21 +410,21 @@ def recast(self, typ=None):
                     # Instead, we raise an error when the object is saved.
                     return
             else:
-                self.type = self._typedmodels_type
+                self.dtm_type = self._typedmodels_type
 
         if typ is None:
-            typ = self.type
+            typ = self.dtm_type
```
Comment (craigds) [inline typedmodels/models.py:416]: Again I think this does a db query, unless it was just set above.

Is that true? I can't find anything suggesting that Django uses the cache for foreignkeys to ContentType.

```suggestion
            typ = ContentType.objects.get_for_id(self.dtm_type_id)```

**Review by craigds — CHANGES_REQUESTED** [review]
Thanks for tackling this!

Code looks great; see comments for suggested improvements

we'll also need: 

* [ ] a changelog entry (0.12)
* [ ] some way to generate migrations for each affected app. I don't _think_ Django has much to help here; we might just have to add a template migration and add it into the release notes. To research.

The migration is actually going to be a bit complex unfortunately... Standard practice is to do this sort of migration in 2-3 steps to avoid downtime:

🅰️ add a nullable field & populate it
🅱️ drop the old field (only run this after switching to the new code)

(If users don't care about downtime they can just run migrations all at once)

But I don't know if that can be communicated to library users clearly.

I wonder if the best way to communicate this migration might be to release a 0.12 that:
* adds the new field but also keeps the `type` field (but only uses it if `dtm_type_id` is not set)
* produces a DeprecationWarning if a model instance is loaded that has a populated `type` but a null `dtm_type_id`, which links to CHANGES.md with a note on how to write the migration.

Then (probably much later, to minimise the number of users skipping 0.12) release another version which drops the old field.

Does that sound sensible?

File: typedmodels/models.py (hunk 4 above)
Comment (IO5) [reply typedmodels/models.py:325]: ah my bad, artefact of the migration, should be set to False after removing the 'type' field

**Comment by IO5** [comment]
> 🅰️ add a nullable field & populate it
> 🅱️ drop the old field (only run this after switching to the new code)

That's what I did basically. And considering I never even edited a migration before, it was surprisingly straightforward. 

As for the handling of the library user, I'm afraid that I won't be of much help here.

File: typedmodels/models.py (hunk 2 above)
Comment (IO5) [reply typedmodels/models.py]: Huh, totally missed that. 

I've applied all of the suggestions, with one difference: as I understand `for_concrete_model=False` is also needed because of proxy the models.

**Comment by craigds** [comment]
I've been thinking about this PR. Your implementation looks solid, thanks for the work you've put into this.

However, I'm starting to think it isn't a good idea ([I'm aware this was my idea :/](https://github.com/craigds/django-typed-models/issues/56#issuecomment-819069825))

It's a fairly major backwards incompatible change, where the main benefit is kind of hard to quantify (it _feels_ right to use a ContentType FK, but doesn't improve things concretely as far as I can tell)

That'd be _okay_ except the migration process looks really far from straightforward - we can't generate useful migrations for the app users; the best we can do is write some instructinos in the template and hope users follow them well. If they fail to do so, it'll probably result in data loss (`django-admin makemigrations` will just add a field and drop another field without copying the types over)

We could mitigate that by giving people a long period to upgrade between 0.11 and 0.12 before dropping the field in 0.13, but they'd still have to follow the release notes, and not everyone's great at following release notes.

So maybe I've put you on the wrong track, not having thought through the migration process when I suggested we should use a ContentType FK, and I'm sorry about that.

I think a simpler change to renaming the field from `type` to `dtm_type` (as originally suggested in #56) with no other field changes would probably work better, because Django detects renamed fields as part of `makemigrations`.

Does that make sense? Once again, thank you for the time you've put into this, I appreciate it

**Comment by IO5** [comment]
Well, one way to solve it is to extract the parts dealing with the type field into a mixin, and then to have a two implementations: legacy one with `CharField` and the new one with `ForeginKey`. 
Then, during the library update, make the user pick one explicitly.

If you like the idea, I can look into this.

**Comment by craigds** [comment]
I think it complicates things a lot and isn't worth it. Let's go with a simple field rename without changing to a ContentType FK instead.

Let me reiterate that I appreciate you taking the time to put this PR together, I just don't think the benefit is worth the cost. Thanks
